scripts/MTweedieTests_parallel.r

```diff
@@ -0,0 +1,156 @@
+# Assignment 1:  
+library(tweedie) 
+library(ggplot2)
+library(parallel)
+library(doParallel)
+
+simTweedieTest <-  
+  function(N){ 
+    t.test( 
+      rtweedie(N, mu=10000, phi=100, power=1.9), 
+      mu=10000 
+    )$p.value 
+  } 
+
+
+# Assignment 2:  
+MTweedieTests <- function(N, M, sig) { 
+  maxcores <- 8
+  Cores <- min(parallel::detectCores(), maxcores)
+  cl <- makeCluster(Cores)
+  registerDoParallel(cl)
+  
+  # Export the simTweedieTest function to each worker
+  clusterExport(cl, varlist = c("simTweedieTest"))
+  
+  p_values <- foreach(i=1:M, .combine='c', .packages='tweedie') %dopar% {
+    simTweedieTest(N)
+  }
+  
+  stopCluster(cl)
+  sum(p_values < sig) / M
+}
+
+
+# Assignment 3:  
+df <-  
+  expand.grid( 
+    N = c(10,100,1000,5000, 10000), 
+    M = 1000, 
+    share_reject = NA) 
+
+
+for(i in 1:nrow(df)){ 
+  df$share_reject[i] <-  
+    MTweedieTests( 
+      N=df$N[i], 
+      M=df$M[i], 
+      sig=.05) 
+} 
+
+##Comment on findings
+
+# After running the timer.r script, I found that the "MTweedieTests Parallel" method was the fastest,
+# completing the task in 27.28. This improvement in speed can be attributed to the parallelization
+# of the M simulations in the MTweedieTests function, allowing the workload to be distributed across 
+# multiple cores. This approach is more efficient than parallelizing the final loop since the
+# MTweedieTests function involves a larger number of independent simulations, providing more opportunities
+# for performance gains through parallel computing. The original script, without any parallelization,
+# was the slowest, highlighting the benefits of leveraging parallel computing in R for computationally
+# intensive tasks.
```

Comment:
The reason that the last one is quicker than the second one is because the second one parallelizes the rows in df. Because there are only 5 rows, having more than 5 cores will not make it quicker. The last one parallizes M, and can therefore utilize more cores in the computation. For example, if you use 10 cores, each core will run M/10 simulations each.